yt_dlp/extractor/bilibili.py

```diff
@@ -644,15 +644,19 @@ def _real_extract(self, url):
             video_data = initial_state['videoInfo']
         else:
             play_info_obj = self._search_json(
-                r'window\.__playinfo__\s*=', webpage, 'play info', video_id, fatal=False)
+                r'window\.__playinfo__\s*=', webpage, 'play info', video_id, default=None)
+            play_info = None
             if not play_info_obj:
                 if traverse_obj(initial_state, ('error', 'trueCode')) == -403:
                     self.raise_login_required()
                 if traverse_obj(initial_state, ('error', 'trueCode')) == -404:
                     raise ExtractorError(
                         'This video may be deleted or geo-restricted. '
                         'You might want to try a VPN or a proxy server (with --proxy)', expected=True)
-            play_info = traverse_obj(play_info_obj, ('data', {dict}))
+                play_info = self._download_playinfo(initial_state['videoData']['bvid'],
+                                                    initial_state['cid'], headers=headers)
+            else:
+                play_info = traverse_obj(play_info_obj, ('data', {dict}))
```

Comment:
if the `playinfo` JSON object has been removed, let's just simplify the code and always make this API call instead

Reply:
not sure if the site will add it back in the future because I just noticed that some of the site's js related to `window.__playinfo__` hasn't been removed yet
```js
if (window.__playinfo__) { primarySetting.prefetch = { playUrl: window.__playinfo__ } }
```

Reply:
With the latest code, the fallback doesn't make the code too complex. You could apply this patch if you still wish to remove it:
```diff
diff --git a/yt_dlp/extractor/bilibili.py b/yt_dlp/extractor/bilibili.py
index 6f798e064..b6bec0827 100644
--- a/yt_dlp/extractor/bilibili.py
+++ b/yt_dlp/extractor/bilibili.py
@@ -680,12 +680,7 @@ def _real_extract(self, url):
         old_video_id = format_field(aid, None, f'%s_part{part_id or 1}')
         cid = traverse_obj(video_data, ('pages', part_id - 1, 'cid')) if part_id else video_data.get('cid')
 
-        play_info = (
-            traverse_obj(
-                self._search_json(
-                    r'window\.__playinfo__\s*=', webpage, 'play info', video_id, default=None),
-                ('data', {dict}))
-            or self._download_playinfo(video_id, cid, headers=headers, fatal=False))
+        play_info = self._download_playinfo(video_id, cid, headers=headers, fatal=False)
         if not play_info:
             raise ExtractorError('Failed to extract play info')
 
 ```